User profiles & redesign of Take Action (action carousel)

manue1 opened the pull request:
**User Profile Task**
https://trello.com/c/ceTtzwrA/96-design-and-add-new-profile-pages

_Route_: `/en/profile/%user_id%`
_Linked to from_: Footprint -> Leaders -> Name link and also after having logged in: Settings -> Profile -> View (links to own profile)

Smoke test:

1. Run API from [features/user-profile](https://github.com/arbolista-dev/cc-user-api/pull/9) branch
2. Sign up new user
3. Go to 'Get Started' and update any value to have him listed as a leader
4. Go to 'Your Footprint' -> Click 'See how you compare' to show leaders and find your full name linking to its profile
5. Click name and find footprint graphs and summary
6. Go to 'Settings -> Find 'Profile' section and click 'View' to also see own user profile
7. Upload a profile photo
8. Add social media links and introduction
9. Make sure photo is uploaded and can be replaced by re-uploading a new one (hover over existing one)
10. Make sure Footprint is correct and bar / chart graphs are displayed correctly
11. Go to Take Action, pledge any action and go back to profile to make sure it shows up in the bottom table.
12. Actions listed in this table can be set to Complete by clicking the tick button. Actions are sorted by status, first completed ones, followed by pledged ones.
13. When in profile copy the URL, open in private browser window to make sure it's accessible without being authenticated (only if set to Public)
14. In authenticated window test 'My profile is public' checkbox at top, untick and re-open profile in private browser window should display error: This user profile is private.
15. Opening non-existing profile: /profile/9999 should display error: The user profile you're trying to access does not exist.

**Action Carousel Tasks**
https://trello.com/c/JWXmpfTA/113-ui-implement-action-carousel
https://trello.com/c/jWCdrZIp/114-ui-implement-action-filtering-and-sorting

_Route_: `/en/take_action/%action_key%`
Action key in URL is optional, if given carousel shows its card and resets the filters to make sure it's not excluded by them.

Smoke test:

1. Run API from [features/user-profile](https://github.com/arbolista-dev/cc-user-api/pull/9) branch
2. Go to 'Take Action' and see More efficient vehicle card, test carousel using arrow buttons
3. Go to '/en/take_action/go_organic' to check routing to specific card works
4. Test Category and Status filters, set some actions as Pledged
5. Test 'Mark this action as not relevant' button and then use filters to make sure it's not showing up within Relevant.
6. Test 'Sort by' functionality, sidenote: Sort by is reset when updating the filters
7. When being logged in as a registered user, the 'Complete' status is available in the status filter and when an action has been pledged
8. Pledging / Completing any actions updates the table in the user profile when being logged in.

Commit:
Add Profile component. Gets user info from User API (/profile/:id).

Commit:
Add Profile component. Gets user info from User API (/profile/:id).

Commit:
Show graph within profile component.

Commit:
Show graph within profile component.

Commit:
Link to Profile in Settings & Footprint components

Show error message for non-existent profile.

Commit:
Link to Profile in Settings & Footprint components

Show error message for non-existent profile.

Commit:
User profile: Add ability to upload photo.

Only editable for own profile when logged in.
After having set photo, upload functionality is shown on hover.

Commit:
User profile: Add ability to upload photo.

Only editable for own profile when logged in.
After having set photo, upload functionality is shown on hover.

Commit:
Add social media profile and introduction to profile.

Commit:
Add social media profile and introduction to profile.

Commit:
Redesign actions as cards within carousel.

Arbolista commented:
Looking great. Exciting, I will preview with Chris tomorrow.

Unas cosillas:

1. Hay unos conflictos con develop.
2. It seems states will show up twice if users have different locations but same state (I am registered in the attachment at zipcodes 93446 and 93465).
3. It'd be great if social media links open in another window.
4. Household size appears as zero for average size households.
5. The upload API calls takes a while. We should put a spinner and disable the button (so user doesn't repeatedly hit button).
6. It'd be preferable if there was a place on the edit profile page to change the privacy setting for the user. How complicated is that?

<img width="1440" alt="screen shot 2016-12-13 at 11 46 49 pm" src="https://cloud.githubusercontent.com/assets/3526083/21170956/78c07e88-c18e-11e6-9289-1a82eb8a7773.png">

Commit:
Merge branch 'develop' into feature/user-profile

Conflicts:
	client/api/real/user.api.js
	shared/components/common/graphs/graphs.component.js
	shared/components/pages/settings/settings.component.js
	shared/lib/base_classes/panel.js
	shared/lib/router/router.js
	shared/lib/state_manager/state_manager.js
	shared/reducers/auth/auth.reducers.js

Commit:
Leaders and user profile fixes.

- Leaders listing: Show each state only once within filter dropdown menu
- Leaders listing: Display average household size as 2.5
- Profile: Show spinner while uploading photo / updating
- Let user update privacy setting from within profile
- Show error if profile set to private

Commit:
Leaders and user profile fixes.

- Leaders listing: Show each state only once within filter dropdown menu
- Leaders listing: Display average household size as 2.5
- Profile: Show spinner while uploading photo / updating
- Let user update privacy setting from within profile
- Show error if profile set to private

manue1 commented:
@Arbolista I have fixed all of the mentioned issues, including the option to update the privacy setting from within the user profile.

For testing please also pull the latest API changes.

Commit:
Add updateActionStatus reducer.Add action filters.

- Rebases changes from develop branch
- Changes the way Take Action components retrieves list of actions to be
shown
- Add Pledge|Complete button pair to cards
- Remodel action filters, by default no category is selected

Commit:
Make sure category filter shows only respective action cards.

Commit:
Store actions in localStorage. Add action status filter.

Commit:
Add sorting of actions.

They can be sorted by either:
- Title
- Tons saved
- Dollars saved
- Down payment

Implementation currently contains rt-if for comparing the sort_by state
property to each of the available options. Without doing this check the
action component does not get re-rendered in the sorted order.

Commit:
Toggle filter and sort options. Don't show by default.

Commit:
Filter by status + category concurrently.

Commit:
Combine filters with sort by.

Changing the filter resets any previously selected sort by option.
This is due to the component not updating if category or status filters
are changed.

A dirty fix would be adding all possible combinations as rt-if's,
however for now I consider resetting the sort by selection as a better
solution.

Commit:
Allow marking unrelevant action as relevant.

Commit:
Show 'Complete' button and status filter only if user = authenticated.

Commit:
Eslint and clean up.

Commit:
Merge branch 'feature/user-profile' of github.com:arbolista-dev/cc-calculator-ui into feature/user-profile

Conflicts:
	client/api/real/user.api.js
	shared/components/pages/profile/profile.component.js
	shared/components/pages/profile/profile.test.js
	shared/lib/base_classes/panel.js
	shared/lib/state_manager/state_manager.js
	shared/reducers/auth/auth.reducers.js
	shared/reducers/profile/profile.reducers.js

Commit:
Changes to updating action and available statuses.

- Action details (tons_saved, dollars_saved, upfront_cost) are floating
point numbers when being updated through User API.
- Show 'Complete' status option in filter and action card only if user
is authenticated.

Commit:
Add user goals table to profile.

Commit:
Update style of Profile user goals table.

Commit:
Simplify retrieving action content. Shopping->Food

Simplify retrieval of content for actions by removing categories.
Rename shopping category to food as its actions are food-related.

Commit:
Simplify passing action status as parameter.

Commit:
Move TA filter and sort to Redux UI state.

Commit:
Fix tests.

Commit:
Serve favicon in production.

Commit:
Each action card has it's own URL.

Commit:
Update pathname when sliding carousel card to reflect current action key.

Commit:
Fix profile issues

- Correctly convert public parameter to string
- Show empty goals table if none have been pledged
- Show introduction after having been set in public mode

erichulburd commented on client/api/test/user.api.js:
```diff
@@ -71,6 +87,14 @@ function setLocation() {
 
+function setPhoto() {
+  return new Promise((fnResolve) => {
+    fnResolve({
```
Creo que solo necesitas `Promise.resolve({ token });`.

erichulburd commented on client/api/test/user.api.js:
```diff
@@ -87,12 +111,60 @@ function listLocations() {
+function showProfile() {
+  return new Promise((fnResolve) => {
+    fnResolve({
+      success: true,
```
Igual aqui `Promise.resolve({ ... });`

erichulburd commented on shared/components/pages/profile/profile.test.js:
```diff
@@ -0,0 +1,28 @@
+  afterEach(() => {
+    const elem = document.getElementById('test');
+    elem.parentNode.removeChild(elem);
+  });
```
👍

erichulburd commented on shared/components/pages/take_action/take_action.component.js:
```diff
@@ -1,44 +1,39 @@
-/* global module*/
+/* global module window*/
 
```
I think in eslint config you can provide 

```
{
  "env": {"browser": true, "node": true}
}
```
 to avoid declaring globals at top.

erichulburd commented on shared/components/pages/take_action/take_action.component.js:
```diff
@@ -1,44 +1,39 @@
+const STATUS_OPTIONS = [{ title: 'All', key: 'all' }, { title: 'Pledged', key: 'pledged' }, { title: 'Not pledged', key: 'not_pledged' }, { title: 'Relevant', key: 'relevant' }, { title: 'Not relevant', key: 'not_relevant' }, { title: 'Completed', key: 'completed' }];
 
+const SORT_OPTIONS = [{ title: ' - ', id: 'inactive' }, { title: 'Title', id: 'title' }, { title: 'Carbon savings', id: 'tons_saved' }, { title: 'Money savings', id: 'dollars_saved' }, { title: 'Down payment', id: 'upfront_cost' }];
 
```
It's likely we will eventually internationalize these.

erichulburd commented on shared/components/pages/take_action/take_action.component.js:
```diff
@@ -1,44 +1,39 @@
+
+  componentWillUnmount() {
+    window.jQuery('#action-carousel').off('slid.bs.carousel');
   }
```
👍

erichulburd commented on shared/components/pages/take_action/take_action.component.js:
```diff
@@ -96,26 +159,166 @@ class TakeActionComponent extends Panel {
+    } else if (status === 'not_pledged') {
+      const pledged = Object.keys(take_action.actions_profile.get('pledged').toJS());
+      const not_pledged = actions_to_filter.filter(key => !pledged.includes(key));
+
```
Avoid casting to JS when you can - it makes easier reading and less time figuring out whether we are dealing with Immutable or plain JS: 

https://facebook.github.io/immutable-js/docs/#/Map/filter

erichulburd commented on shared/containers/profile.container.js:
```diff
@@ -0,0 +1,41 @@
+    updateUI.assignTo(dispatch);
+    updateUI(payload);
+  },
+});
```
I originally made a mistake using assignTo. All you really need here is `dispatch(updateUI(payload))`. 

There can be some funky side effects to using assignTo, but don't worry about changing this until we actually discover a specific issue related to this.

erichulburd commented on shared/reducers/profile/profile.reducers.js:
```diff
@@ -0,0 +1,59 @@
+    loading: <Boolean>,
+    load_error: <Boolean>
+  }
+*/
```
👍

erichulburd reviewed (commented):
Code overall looks good. None of these comments are critical, just some things to keep in mind.

Arbolista commented:
After logging in, locally I receive: `GET http://localhost:8082/profile/1 404 (Not Found)`.

We can review together Monday morning or after it's been fixed.

Commit:
Internationalize Take Action filters/sort by.

manue1 commented:
> After logging in, locally I receive: GET http://localhost:8082/profile/1 404 (Not Found).

I couldn't reproduce this error. Not really sure why you get an 404 on that URL, because `/profile/:userId` is the path structure of the UI, but the UI is running on port 3000.
I changed the API route to `/user/:userId/profile` so I'm not sure if you're running UI & API up-to-date with the latest commits that I've pushed?

Commit:
Fix displaying amount of savings in sidebar.

Commit:
Optimize mobile view of TA carousel.
